mne/simulation/source.py

```diff
@@ -312,3 +317,188 @@ def simulate_stc(src, labels, stc_data, tmin, tstep, value_fun=None,
     stc = SourceEstimate(np.concatenate(data), vertices=vertno, tmin=tmin,
                          tstep=tstep, subject=subject)
     return stc
+
+
+class SourceSimulator():
+    """Class to generate simulated Source Estimates.
+
+    Parameters
+    ----------
+    src : instance of SourceSpaces
+        Source space.
+    tstep : float
+        Time step between successive samples in data. Default is 0.001 sec.
+    duration : float | None
+        Time interval during which the simulation takes place in seconds.
+        Default value is computed using existing events and waveform lengths.
+
+    Attributes
+    ----------
+    duration : float
+        The duration of the simulation in seconds.
+    """
+
+    def __init__(self, src, tstep=1e-3, duration=None):
+        self._src = src
+        self._tstep = tstep
+        self._labels = []
+        self._waveforms = []
+        self._events = np.empty((0, 3), dtype=int)
+        self._duration = duration
+        self._last_samples = []
+        self._chk_duration = 1000
+
+    @property
+    def duration(self):
+        """Duration of the simulation"""
+        # If not, the precomputed maximum last sample is used
+        if self._duration is None:
+            return np.max(self._last_samples) * self._tstep
+        return self._duration
+
+    @property
+    def nb_samples(self):
+        """Number of samples in the simulation"""
+        return int(self.duration / self._tstep)
+
+    def add_data(self, label, waveform, events):
+        """Add data to the simulation
+
+        Data should be added in the form of a triplet of
+        Label (Where) - Waveform(s) (What) - Event(s) (When)
+
+        Parameters
+        ----------
+        label : Label
+            The label (as created for example by mne.read_label). If the label
+            does not match any sources in the SourceEstimate, a ValueError is
+            raised.
+        waveform : list | array
+            The waveform(s) describing the activity on the label vertices.
+            If list, must have the same length as events
+        events: array of int, shape (n_events, 3)
+            Events associated to the waveform(s) to specify when the activity
+            should occur.
+        """
+        if not isinstance(label, Label):
+            raise ValueError('label must be a Label,'
+                             'not %s' % type(label))
+
+        # If it is not a list then make it one
+        if not isinstance(waveform, list) or len(waveform) == 1:
+            waveform = [waveform] * len(events)
+            # The length is either equal to the length of events, or 1
+        if len(waveform) != len(events):
+            raise ValueError('Number of waveforms and events should match '
+                             'or there should be a single waveform')
```

Comment:
say what values you got. Example

https://github.com/mne-tools/mne-python/blob/master/mne/event.py#L866